### ISSUE TITLE
QS: parse and render ELF structures

### ISSUE THREAD

**Issue by williballenthin** [mandiant/flare-floss#785]

**Comment by c-urly** [comment]
Hey, Is anybody working on this?
I was thinking of making floss supportable for the ELF file. I kind of got the gist of the core flow after browsing the code, will try to add support for ELF. Also, is there a dev chat forum? it will be easier to ask questions if I am stuck somewhere.

If someone already has a layout in mind to implement this, please do share.

**Comment by FredCoast** [comment]
I've made a start on this, I believe I have successfully gotten the headings/reloc and sections and applied it to the output,

However, I am struggling to understand what lancelot does for the PE files and how I can apply that to ELF files, if you have any pointers?

**Comment by williballenthin** [comment]
lancelot is a disassembler library that we use here to recover code and functions. unfortunately, its undergone some revisions since QS started using it, and furthermore, I don't expect to be able to develop it further soon.

it might be worth migrating this project to use idalib instead. note that i may be biased these days, though.

**Comment by FredCoast** [comment]
Sounds good I'll look into it, would that also support mach-o files too?

And any tips on where to start?

**Comment by williballenthin** [comment]
yes, IDA supports mach-o. though, to parse the file structure we should use a Python library (maybe macholib).

do you have an IDA license?

**Comment by FredCoast** [comment]
I'll use macholib for that!

No I don't, I searched it up after replying and quickly realised, sorry about that

**Comment by FredCoast** [comment]
I was thinking, what if I created my own version of Lancelot for elf files, it would largely be as a learning experience for myself more than anything,

I thought if I used goblin and capstone-rs to cover four main areas:
1. Symbol tables
2. Relocations
3. Disassembly/Stepping
4. Heuristics

With potentially expansion for signatures, would that work in addition of Lancelot for this scenario?

**Comment by williballenthin** [comment]
if you're willing to dive into Rust code, then an alternative is to update QS to use the new version of Lancelot (i had assumed most people wouldn't be interested in this). the new API is much simpler: input bytes and receive a BinExport2 structure of the results. 

your idea is fine too, though it will be a bit of work (also fine!) so we should just be realistic and honest about that.

**Comment by williballenthin** [comment]
you could update Lancelot to support ELF files, too. that's not hard, just wasn't my focus at the time. already the loader is abstracted over PE and COFF so adding a new format shouldn't be too disruptive.

**Comment by FredCoast** [comment]
Yeah I would be happy to look into it, I'd just assumed that it would take the same amount of work to update Lancelot and create a new app so that's why I suggested it. 

By updating to the latest version I believe that it would fix some other issues too.

I will look into it, Thanks for the suggestions!

**Comment by FredCoast** [comment]
I have a question about stripped elf files and I was wondering if you could help me with it,

So they lose their offsets and names etc, how would you recommend that I go about handling them?

Would getting as much information from the program header be the best option despite a theoretical section potentially containing multiple sections?

**Comment by williballenthin** [comment]
i think thats fine. we should be trying to do our best, recognizing that often some data may not be available.

**Comment by FredCoast** [comment]
I believe I have implemented it correctly, I've created a fork if you wanted to check it out,

There are some points for improvement but the basic functionality is implemented but I will continue working on it when I can

Also the last comment I was getting sections and segments mixed up that's my bad

**Comment by williballenthin** [comment]
thats awesome! want to open a PR and we can discuss it there?

**Comment by FredCoast** [comment]
Hopefully, I have an implementation for this now 

I need to handle the import tables (and lancelot) but it renders the structure of the elf file and labels "structure headers" for relevant strings.

I can send it over if you would like!

Double checking with the import tables is it highlighting that that string is the name of the imported library/dll?

**Comment by williballenthin** [comment]
exactly. ideally users should be able to see that a string is related to the file structure (eg the import table) without being familiar with the typical layout/pattern of an ELF file. 

we'd want the structure column in QS to be filled. and, if the strings for the import table of an ELF file are typically stored discretely, then we might want a delimited section in the QS output (however, I seem to recall the import strings are found in a string table with other strings in an ELF file, so this probably won't be possible).

**Comment by williballenthin** [comment]
more generally, when reviewing the output of readelf and cross referencing against the strings from the file, ideally QS should "explain" as many strings entries as possible. so strings related to notes, symbols, sections, imports, etc should all be marked up and/or delimited.

**Comment by FredCoast** [comment]
I'll keep that in mind, I can add some headers which I think are appropriate and would you be able to double check them if you think they are sensible please

> I seem to recall the import strings are found in a string table with other strings in an ELF file, so this probably won't be possible

There is a way I believe, through these things called 'DT_NEEDED' in the '.dynamic' which points to an offset which contains the name. Another way if I am wrong is another tagging database could be used for common library names?

**Comment by williballenthin** [comment]
i think it should be possible to tag the strings as from the import table. but i don't think it will be likely that you can subdivide the strings table into the strings related to imports (since we can't assume all the import related string entries will be together)

**Comment by FredCoast** [comment]
Ah sorry I wasn't clear in what I was saying. I was meaning getting the names of the libraries, so it would look like this potentially:

<img width="534" height="73" alt="Image" src="https://github.com/user-attachments/assets/f82f2211-55d0-42e9-9b74-f829553d8dca" />

(I notice the GLIBC but I don't think there is anything I can do about it)

I agree with the fact that the strings themselves probably couldn't be split up, sorry for the confusion.
